### PR TITLE
[FIX] mail: translate group-specific items

### PR DESCRIPTION
**Pull request by reth-odoo**
Currently all group-specific buttons are generated within the same language context.

This means if I send a message to someone who speaks dutch but I speak french, the recipient will receive action buttons in french when we could have detected that they should have received it in dutch.

task-3370576

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/models/mail_thread.py [hunk 1]
```diff
@@ -2444,7 +2443,9 @@ def _notify_prepare_template_context(self, message, msg_vals, model_description=
             if template and template.lang:
                 lang = template._render_lang([self.env.context['default_res_id']])[self.env.context['default_res_id']]
 
-        if not model_description and model:
+        if model_id:
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:2446]: Not sure to understand this change ... I suppose the issue is that we propagate an already-translated model_description ?

File: addons/mail/models/mail_thread.py [hunk 2]
```diff
@@ -2629,37 +2630,51 @@ def _notify_classify_recipients(self, recipient_data, model_name, msg_vals=None)
         }]
         only return groups with recipients
         """
+        recipient_lang_groups = self.env['res.partner'].sudo().read_group(
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:2633]: Side note: "groups" suffix is misleading, as we are already dealing with "groups" (using notify meaning of groups).

File: addons/mail/models/mail_thread.py [hunk 3]
```diff
@@ -2629,37 +2630,51 @@ def _notify_classify_recipients(self, recipient_data, model_name, msg_vals=None)
...
-            group_data.setdefault('notification_group_name', group_name)
-            group_data.setdefault('notification_is_customer', False)
-            is_thread_notification = self._notify_get_recipients_thread_info(msg_vals=msg_vals)['is_thread_notification']
-            group_data.setdefault('has_button_access', is_thread_notification)
-            group_button_access = group_data.setdefault('button_access', {})
-            group_button_access.setdefault('url', access_link)
-            group_button_access.setdefault('title', view_title)
-            group_data.setdefault('actions', list())
-            group_data.setdefault('recipients', list())
+        for lang, lang_groups in groups_by_lang.items():
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:2648]: So I suppose this is the "real" fix (a bit like what is done in _notify_get_classified_recipients_iterator in newer versions), looping on langs then each lang has recipients divided into groups (user, portal, ...) ?

File: addons/mail/models/mail_thread.py [hunk 4]
```diff
@@ -2629,37 +2630,51 @@ def _notify_classify_recipients(self, recipient_data, model_name, msg_vals=None)
...
+        for lang, recipient_ids in [(group['lang'], group['ids']) for group in recipient_lang_groups]:
+            for recipient_id in recipient_ids:
+                for group_name, group_func, group_data in groups_by_lang[lang]:
+                    if group_func(partner_id_to_recipient_data[recipient_id]):
+                        group_data['recipients'].append(recipient_id)
+                        break
 
         result = []
-        for group_name, group_method, group_data in groups:
+        for group_name, dummy, group_data in [
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:2676]: If I am not wrong, this changes the returned value of "_notify_classify_recipients", and groups are not unique anymore ... which is annoying in stable (and not aligned with master I think, where we first classify by lang then classify by group if I am not wrong)

File: addons/mail/models/mail_thread.py [hunk 5]
```diff
@@ -2598,7 +2599,7 @@ def _notify_get_groups(self, msg_vals=None):
             )
         ]
 
-    def _notify_classify_recipients(self, recipient_data, model_name, msg_vals=None):
+    def _notify_classify_recipients(self, recipient_data, model_name, msg_vals=None, model_id=None):
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:2602]: Same issue with parameter here ^^

**Review by tde-banana-odoo — COMMENTED** [review]
Some comments, but mainly "not sure we should do that in stable" comments :D

**Review by reth-odoo — COMMENTED** [review]
Agreed with the comments, I'd say that was the least breaking option as otherwise we have to override the result of `_notify_prepare_template_context` which just won't make sense.

Adding groups was really the only way to keep the same return type so if we consider it's not ok (and I can see why), I'd say it's not "fixable" in stable then.

**Comment by tde-banana-odoo** [comment]
@reth-odoo Would say: not fixable in stable, especially it is globally fixed in 17. Moreover it is a complicated change (input / output, potential regressions, custom overrides, ...) so it is clearly not stable policy compliant. I agree it is a bit annoying in stable but it is not preventing people from working (no traceback, no issue), ... it is more a "comfort" fix than a real bugfix.

All in all, I think it is better to not fix it in stable. Closing then !

**Pull request closed without merging**